pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,474 @@
+	}
+
+	if err := m.updateLatencyConfig(newNLM); err != nil {
+		klog.V(4).ErrorS(err, "Failed to update latency config")
+	}
+}
+
+func (m *NodeLatencyMonitor) updateLatencyConfig(nlm *v1alpha1.NodeLatencyMonitor) error {
+	// Parse the ping interval
+	pingInterval := time.Duration(nlm.Spec.PingIntervalSeconds) * time.Second
+
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{
+		Enable:   true,
+		Interval: pingInterval,
+	}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+
+	return nil
+}
+
+// onNodeLatencyMonitorDelete is the event handler for deleting NodeLatencyMonitor.
+func (m *NodeLatencyMonitor) onNodeLatencyMonitorDelete(obj interface{}) {
+	// Update the latency config
+	m.latencyConfig = &LatencyConfig{Enable: false}
+
+	// Notify the latency config changed
+	m.latencyConfigChanged <- struct{}{}
+}
+
+func (m *NodeLatencyMonitor) sendPing(socket net.PacketConn, addr net.IP) error {
+	var requestType icmp.Type
+
+	// Resolve the IP address
+	ip := &net.IPAddr{IP: addr}
+
+	// Create a new ICMP packet
+	if addr.To4() == nil {
+		requestType = ipv6.ICMPTypeEchoRequest
+	} else {
+		requestType = ipv4.ICMPTypeEcho
+	}
+
+	timeStart := time.Now()
+	seqID := getICMPSeq()
+	body := &icmp.Echo{
+		ID:   os.Getpid() & 0xffff,                       // Use the current process ID as the ICMP ID
+		Seq:  int(seqID),                                 // Use the current sequence number as the ICMP sequence number
+		Data: []byte(timeStart.Format(time.RFC3339Nano)), // Store the current time in the ICMP data
+	}
+	msg := icmp.Message{
+		Type: requestType,
+		Code: 0,
+		Body: body,
+	}
+	klog.V(4).InfoS("Send ICMP message", "IP", ip, "SeqID", seqID, "body", body)
+
+	// Serialize the ICMP message
+	msgBytes, err := msg.Marshal(nil)
+	if err != nil {
+		return err
+	}
+
+	// Send the ICMP message
+	_, err = socket.WriteTo(msgBytes, ip)
+	if err != nil {
+		return err
+	}
+
+	// Create or update the latency store
+	mutator := func(entry *NodeIPLatencyEntry) {
+		entry.LastSendTime = timeStart
+		entry.SeqID = seqID
+	}
+	m.latencyStore.SetNodeIPLatencyEntry(addr.String(), mutator)
+
+	return nil
+}
+
+func (m *NodeLatencyMonitor) recvPing(socket net.PacketConn, isIPv4 bool, stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			return
+		default:
+			// max size of the ICMP message is 1500 bytes, which is the maximum size of an Ethernet frame.
+			readBuffer := make([]byte, 1500)
+			n, peer, err := socket.ReadFrom(readBuffer)
+			if err != nil {
+				if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+					klog.V(4).ErrorS(err, "Timeout reading ICMP message")
+					continue
+				}
+				klog.V(4).ErrorS(err, "Failed to read ICMP message")
+			}
+
+			destIP := peer.String()
+			// Get the node name by destIP
+			entry, ok := m.latencyStore.GetNodeIPLatencyEntry(destIP)
+			if !ok {
+				klog.V(4).InfoS("Failed to get node entry by destIP", "destIP", destIP)
+				continue
+			}
```

Comment:
when my suggestion of removing `SeqID` from `NodeIPLatencyEntry` is implemented, this can be removed completely.
I don't like the fact that we do a GET here today, because it feels like there is a possibility for someone (another goroutine) to come and update the entry in the store, while we process the reply below.

Reply:
Ok, here I have removed.

BTW, the `GET` method here has been modified to return a copy instead of a pointer, which can't be modified directly.